pkg/controller/master/node/promote_controller_test.go

```diff
@@ -0,0 +1,331 @@
+package node
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type NodeBuilder struct {
+	node *corev1.Node
+}
+
+func NewDefaultNodeBuilder() *NodeBuilder {
+	return &NodeBuilder{
+		node: &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:              "",
+				Labels:            map[string]string{},
+				Annotations:       map[string]string{},
+				CreationTimestamp: metav1.NewTime(time.Now()),
+			},
+			Status: corev1.NodeStatus{
+				Conditions: []corev1.NodeCondition{},
+			},
+		},
+	}
+}
+
+func (n *NodeBuilder) Name(name string) *NodeBuilder {
+	n.node.Name = name
+	return n
+}
+
+func (n *NodeBuilder) Harvester() *NodeBuilder {
+	n.node.Labels[HarvesterManagedNodeLabelKey] = "true"
+	return n
+}
+
+func (n *NodeBuilder) Master() *corev1.Node {
+	n.node.Labels[KubeMasterNodeLabelKey] = "true"
+	n.node.CreationTimestamp = metav1.NewTime(time.Now())
+	return n.node
+}
+
+func (n *NodeBuilder) Running() *NodeBuilder {
+	n.node.Annotations[HarvesterPromoteStatusAnnotationKey] = PromoteStatusRunning
+	return n
+}
+
+func (n *NodeBuilder) Complete() *NodeBuilder {
+	n.node.Annotations[HarvesterPromoteStatusAnnotationKey] = PromoteStatusComplete
+	return n
+}
+
+func (n *NodeBuilder) Failed() *NodeBuilder {
+	n.node.Annotations[HarvesterPromoteStatusAnnotationKey] = PromoteStatusFailed
+	return n
+}
+
+func (n *NodeBuilder) NotReady() *NodeBuilder {
+	ready := corev1.NodeCondition{
+		Type:   corev1.NodeReady,
+		Status: corev1.ConditionFalse,
+	}
+	n.node.Status.Conditions = append(n.node.Status.Conditions, ready)
+	return n
+}
+
+func (n *NodeBuilder) Node() *corev1.Node {
+	n.node.CreationTimestamp = metav1.NewTime(time.Now())
+	return n.node
+}
+
+var (
+	M1 = NewDefaultNodeBuilder().Name("M1").Master()
```

Comment:
nit: It can work in tests but generally uppercase is not valid in Kubernetes resource names.